src/org/pentaho/reporting/platform/plugin/async/PentahoAsyncReportExecution.java

```diff
@@ -82,9 +82,12 @@ public void notifyTaskQueued( final UUID id ) {
         AuditHelper.audit( safeSession.getId(), safeSession.getId(), url, getClass().getName(), getClass().getName(),
             MessageTypes.INSTANCE_END, auditId, "", ( (float) ( end - start ) / 1000 ), null );
 
+
+        final InputStream stagingContent = handler.getStagingContent();
+
         listener.setStatus( AsyncExecutionStatus.FINISHED );
 
-        return handler.getStagingContent();
+        return stagingContent;
```

Comment:
Even in such way will not completely remove race condition since listener.setStatus(...) is synchronized+volatile - but sycnronized according to a listener class while future completition is in sync with future itself, not according to a listener.